### PR TITLE
8295012: Arena should not derive from CHeapObj<mtNone>

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change to the `Arena` class.  Rather than deriving from
`CHeapObj<mtNone>` (which has the effect of preventing C-heap allocation for an
arena by asserting if such is attempted), instead have no allocation base
class so that `new Arena(...)` will fail to compile.

Testing: mach5 tier1

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295012](https://bugs.openjdk.org/browse/JDK-8295012): Arena should not derive from CHeapObj<mtNone>

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to [828541f5](https://git.openjdk.org/jdk/pull/10621/files/828541f574f2036640bdda37071cf3f9b6443108)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [828541f5](https://git.openjdk.org/jdk/pull/10621/files/828541f574f2036640bdda37071cf3f9b6443108)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10621/head:pull/10621` \
`$ git checkout pull/10621`

Update a local copy of the PR: \
`$ git checkout pull/10621` \
`$ git pull https://git.openjdk.org/jdk pull/10621/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10621`

View PR using the GUI difftool: \
`$ git pr show -t 10621`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10621.diff">https://git.openjdk.org/jdk/pull/10621.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/memory/arena.hpp [hunk 1]
```diff
@@ -86,7 +86,7 @@ class Chunk: CHeapObj<mtChunk> {
 
 //------------------------------Arena------------------------------------------
 // Fast allocation of memory
-class Arena : public CHeapObj<mtNone> {
+class Arena {
```
Comment (dholmes-ora) [inline src/hotspot/share/memory/arena.hpp:89]: The allocation base class tells the reader that `Arena` has special allocation constraints, by removing this it is not apparent there are any such constraints. Can we at least add a comment?

I confess I've forgotten how not having an allocation base class prevents use of `new Foo()` ??

Comment (stefank) [reply src/hotspot/share/memory/arena.hpp:89]: We have a link-time check in debug builds that fails the build if someone tries to use the global new operators.

**Review by stefank — COMMENTED** [review]
What's the reason for keeping the new/delete operators in Arena, instead of inheriting from the newly introduced CHeapObjBase class?

**Comment by kimbarrett** [comment]
> What's the reason for keeping the new/delete operators in Arena, instead of inheriting from the newly introduced CHeapObjBase class?

No good reason.  I just forgot about them.  I will make that change.

**Review by dholmes-ora — APPROVED** [review]
LGTM!

Thanks

**Comment by kimbarrett** [comment]
Thanks @stefank and @dholmes-ora  for reviews.

**Pull request closed without merging**